docs/extension.rst

```diff
@@ -5,46 +5,214 @@ Extension
+words.
 
 Extending the DSL
 *****************
 
-TODO What extending the dsl means and what bits there are
+The DSL words are types, adding a word is as simple as declaring a new type with at least one of the following static
+template methods defined. Each undefined static template method will automatically behave as a no-operation.
+
+The template of each of these static methods should be `template <typename DSL>`, this will have the parsed DSL words
+passed in. It is important to note that the type will only be considered by NUClear in a static context. If any
+attributes need to be stored in the DSL word type template it and use static variables, see `Sync`.
+
+There are DSL words that are not meant to be used directly but as a part of other words, see `CacheGet` and `TypeBind`.
+TODO explain what these two do.
+
+If the type you want to become a DSL extension word is not defined within your control specialise `DSLProxy<>` with the
+type. Provide the template methods to the specialisation of `DSLProxy<>` as if it were the type.
 
 Bind
 ----
 
-TODO What extension via bind means and how it works
+.. codeblock:: c++
+    template <typename DSL>
+    static inline void bind(const std::shared_ptr<threading::Reaction>& reaction, /*More arguments can be declared*/)
+
+This function is called when the reaction is bound, it should be thought of as the constructor. It is used to setup
+anything that is required by the DSL word.
+
+A common use for extensions is to setup something that will generate tasks for this reaction. This can be done by
+communicating to an extension reactor via a helper type that the extension reactor triggers on.
+
+An unbinder, if needed, should be passed to the reaction's unbinders callback list from the bind call. This is used as a
+destructor.
+e.g. for the `IO` word we have
+.. codeblock:: c++
+    reaction->unbinders.push_back([](const threading::Reaction& r) {
+        r.reactor.emit<emit::Direct>(std::make_unique<operation::Unbind<IO>>(r.id));
+    });
+
+which will tell the extension reactor that this reaction no longer exists.
+
+The extra arguments are passed in, in order, from the `on` call.
 
 Get
 ---
 
-TODO what extension via get means and how it works
-Is used to get data to run a function with
-If the type can be dereferenced, then this is also an acceptable type in the argument list
+.. codeblock:: c++
+    template <typename DSL>
+    static inline T get(threading::Reaction&)
+
+This is used to get the data for the callback. The returned value is passed to the callback.
+
+If the return type can be dereferenced, then either the return type or the type returned by the dereference of the
+return type can be used in the callback.
+
+If data needs to be passed to a task when it is submitted to the `Powerplant` use `ThreadStore<T>`. The thread store is
+a static variable that can be accessed from within the get method.
```

Comment:
Might be good to mention you should clear the thread cache afterwards to ensure future invocations won't get stale data